File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.java
class Solution {

    //think like this, if i sell today? how can i get max profit?
    // may be if i sold least before me then -> maintain it,
    // but we expect it to increase? -> so we maintain at each point

    // TC -> O(n)
    // sc -> O(1)
    public int maxProfit(int[] prices) {
        int maxProfit = 0;
        int lastMin = prices[0];
        
       for(int price : prices) {
        maxProfit = Math.max(maxProfit, price - lastMin);
        lastMin = Math.min(lastMin, price);
       } 

       return maxProfit;
    }
}

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.cs
public class Solution {
    public int MaxProfit(int[] prices) {

        int minPrice = int.MaxValue;
        int maxProfit = 0;

        for(int i = 0; i < prices.Length; i++) {
            minPrice = Math.Min(prices[i], minPrice);
            maxProfit = Math.Max(prices[i] - minPrice, maxProfit);
        }

        return maxProfit;
    }
}